fix: make `zig cc` pass `-l/-L` like Clang/GCC for ELF

a-khabarov opened the pull request:
This PR is my attempt to fix #19699 by making the way `zig cc` passes `-l/-L` flags for ELF linking consistent with Clang and GCC.
I've added a test case that mimics the example in the issue description. Without the changes this PR makes to `src`, the new test case fails as expected with
```
========= expected to find: ==========================
NEEDED libfoo.so
========= but parsed file does not contain it: =======
dynamic section
RUNPATH foo
NEEDED foo/libfoo.so
NEEDED libc.so.6
```
I am not sure if `test/tests.zig` is the best place for it - I couldn't find a good way to test this in `test/link/elf.zig` or `test/standalone`.

a-khabarov commented:
@motiejus @kubkon This PR builds on the functionality implemented in https://github.com/ziglang/zig/issues/15743 and should also fix `cc_test` not working correctly when using `hermetic_cc_toolchain` as mentioned in https://github.com/ziglang/zig/issues/15743#issuecomment-2064878213.

kubkon commented on test/tests.zig:
```diff
@@ -769,6 +769,54 @@ pub fn addCliTests(b: *std.Build) *Step {
         step.dependOn(&cleanup.step);
     }
 
+    {
```
This set of tests belongs in linker tests in `test/link/elf.zig`. I also don't understand why can't it be orchestrated using the build system? Why do you need to invoke zig exe via `addSystemCommand`? `exe.linkSystemLibrary("foo")` emits as `-lfoo` fwiw. Did you actually have a look at `test/link/elf.zig`? We do have a substantial amount of tests that rely on emitting and linking against system libs.

a-khabarov replied:
I've moved the testing to `test/link/elf.zig`. The new commits add a few bits of extra functionality that are needed to reproduce the problem.

a-khabarov replied:
@kubkon I think I've addressed your comments on this PR, any thoughts?

a-khabarov replied:
@andrewrk @kubkon Could you please have a look at this PR?

a-khabarov replied:
@andrewrk @kubkon Is there any way we can land this PR in 0.14.0? Please see my email regarding this

a-khabarov commented:
I am in the process of rebasing this on current mainline

a-khabarov commented:
I've rebased this PR.

The fix itself is in `src/link/Elf.zig`.

The new test is in `test/link/elf.zig`. It follows the example from
https://github.com/ziglang/zig/issues/19699.

The rest of the changes are new functionality needed for the test:
* `omit_soname` option for `Build.Module`, `OverlayOptions`.
  This passes `-fno-soname` to the linker when enabled.
* `addLibraryPathSpecial` for `Build.Module`, `Build.Step.Compile`.
  Like `addLibraryPath` but for `[]const u8`.
  `addRPathSpecial` is also added for completeness.
* `setCwd` for `Build.Step.Compile`.
  Allows setting current directory for compile steps.
* `opt_cwd` for `evalZigProcess` in `Build.Step`.
  Allows running `evalZigProcess` in a specific directory.
  Needed for setting current directory for compile steps.
* `--zig-lib-dir` is now appended differently in `Build.Step.Compile`.
  This is needed for compatibility with `setCwd`.

a-khabarov commented:
I made some changes to avoid reconstructing the library name and directory in `src/link/Elf.zig`. 
Everything is now passed via `src/link.zig`

alexrp commented on lib/std/Build/Module.zig:
```diff
@@ -18,6 +18,7 @@ frameworks: std.StringArrayHashMapUnmanaged(LinkFrameworkOptions),
 link_objects: std.ArrayListUnmanaged(LinkObject),
 
 strip: ?bool,
+soname: ?bool,
```
Shouldn't this be using the `SOName` type in `src/main.zig`? (Which can be moved to `std.zig.SoName` to enable sharing it.)

a-khabarov replied:
I've moved `SOName` to `std.zig` and updated this to use it instead of `bool`.

alexrp commented on lib/std/zig.zig:
```diff
@@ -315,6 +315,12 @@ pub const BuildId = union(enum) {
 
 pub const LtoMode = enum { none, full, thin };
 
+pub const SOName = union(enum) {
```
Style nitpick:

```suggestion
pub const SoName = union(enum) {
```

a-khabarov replied:
Changed to `SoName`

alexrp commented on src/link.zig:
```diff
@@ -1818,6 +1818,8 @@ pub const Input = union(enum) {
         needed: bool,
         weak: bool,
         reexport: bool,
+        name: ?[]const u8,
+        lib_directory: Directory,
```
These new fields need to be handled in `hashInputs`.

a-khabarov replied:
I've added these to `hashInputs`

alexrp commented on test/link/elf.zig:
```diff
@@ -82,6 +82,7 @@ pub fn testAll(b: *Build, build_opts: BuildOptions) *Step {
 
         // glibc tests
         elf_step.dependOn(testAsNeeded(b, .{ .target = gnu_target }));
+        elf_step.dependOn(testLibraryPathsCompatibility(b, .{ .target = gnu_target, .use_lld = true }));
```
Bonus points here if this test can be made to pass with `.use_lld = false` too.

a-khabarov replied:
I've added an invocation with `.use_lld = false` and it also passes.

alexrp reviewed (approved):
This seems okay to me. But I'm also not deeply familiar with this stuff, so it's possible that there are subtle details that I'm missing.

@andrewrk do you want to have a look before this gets merged?

andrewrk commented:
Yes, thank you.

birunts commented:
Hi, any updates on this topic? Is this going to be merged as it's already approved?

chrisirhc commented:
Tried to rebase this but looks like `linkWithLLD` has been removed via https://github.com/ziglang/zig/commit/2fb6f5c1adcd764372ad28ed4014fdaf558da778 .

I'm guessing that the refactor that @mlugg is doing will defer this fix PR.

cameron-martin commented:
> No particularly strong reason. I suspected the issues were related but evidently from my attempt, they are not. And also that the workaround in [#19699 (comment)](https://github.com/ziglang/zig/issues/19699#issuecomment-2729192138) is to use `-l:libfoo.so` arguments.

Makes sense.

> In that case, do you think I should clean up #24457 to get a review so that at least the initial issue is resolved when using arguments like `-lmylib`?

Yes please, this would be really useful if you could!

chrisirhc commented:
> > Tried to rebase this as it looks like `linkWithLLD` has been removed via [2fb6f5c](https://github.com/ziglang/zig/commit/2fb6f5c1adcd764372ad28ed4014fdaf558da778) .
> > Attempt to rebase is at #24457 .
> > Update: seems like the link logic is leading to panic("TODO"), that's what I get after applying #24457 and running the reproduction per #23287 on the built binary . Likely since this logic hasn't been implemented since 0.14 , per [5ca5403](https://github.com/ziglang/zig/commit/5ca54036ca0bc292ead681c03c8ac57e27a127db) .
> 
> Not quite sure I follow this - when I pull #24457 and run the reproducer in #19699 it seems to work. Why do you expect this to also fix #23287?

I tested the reproduction in https://github.com/chrisirhc/zig-cc-repro/blob/name-lookup/Makefile using the zig 0.15.2 release, and it just works. So perhaps it's just that the main branch has fixed the name-lookup (use with `-L. -lmylib`) .
I don't think this PR fixes anything with regard to the name-lookup case. 

I was looking for a fix for the file name case `-l :mylib.so`, but looks like there's no fix for that.

I went ahead to update to clean up the PR https://github.com/ziglang/zig/pull/24457 , but I don't see any impact of this PR.
